Feature/285 schools lookup component

tomisaacmarshall opened the pull request:
Type: patch

Partially fixes #285 

## Changes proposed in this PR

This PR partially completes task 285 to recreate the old Storybook schools lookup component in the component library. 

I wanted to break it down so the PR wasn't too big (and because I was having a little trouble with the manual form part), so this includes the school lookup input only - the manual address entry / edit form will come later in a separate PR.

### New components

It adds three components:

- TextInputWithDropdown (atom) - this is a dumb component that just takes care of the visual side of things

- Typeahead (molecule) - this is a generic typeahead component that uses the above, and could be used for various lookup / search input fields. It adds the core functionality around fetching data in a debounced manner, managing / updating state, etc

- SchoolLookup (molecule) - this is the specific schools lookup version of the above component 

### Outstanding issues

- I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!

- Accessibility is another area that I need to improve on. I'm not sure if TextInputWithDropdown  passes muster in this regard - again, grateful for any tips

- Do I need to do any more on styling? The dropdown is based on the original component; I wasn't working from any designs.

gusliedke commented on src/components/Atoms/TextInputWithDropdown/TextInputWithDropdown.style.js:
```diff
@@ -0,0 +1,51 @@
+
+const Dropdown = styled.div`
+  font-family: ${({ theme }) => theme.fontFamilies('Montserrat')};
+  z-index: 1;
```
Add `${zIndex('high')};` from _utils_

gusliedke commented on src/components/Atoms/TextInputWithDropdown/TextInputWithDropdown.style.js:
```diff
@@ -0,0 +1,51 @@
+`;
+
+const DropdownItem = styled.li`
+  line-height: 1.5rem;
```
you can use our _spacing_ `${spacing('md)'}`

AndyEPhipps commented:
> I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!

It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?

gusliedke commented on src/components/Molecules/SchoolLookup/SchoolLookup.test.js:
```diff
@@ -0,0 +1,13 @@
+import renderWithTheme from '../../../hoc/shallowWithTheme';
+import SchoolLookup from './SchoolLookup';
+
+// todo: can we test the interactions?
```
Usually we test interactions in our app. We can add some cypress tests in the integration part.

KrupaPammi commented:
> > I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!
> 
> It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?

I'm working on adding cypress tests on CL @AndyEPhipps

gusliedke commented:
> > I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!
> 
> It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?
> 
> > Accessibility is another area that I need to improve on. I'm not sure if TextInputWithDropdown passes muster in this regard - again, grateful for any tips
> 
> Probably worth installing the Lighthouse browser extension as a starting point

I think we could leave integration test on app level. What do u think?

AndyEPhipps commented:
> > > I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!
> > 
> > 
> > It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?
> > > Accessibility is another area that I need to improve on. I'm not sure if TextInputWithDropdown passes muster in this regard - again, grateful for any tips
> > 
> > 
> > Probably worth installing the Lighthouse browser extension as a starting point
> 
> I think we could leave integration test on app level. What do u think?

I think it's worth it; we've had a few problems end up on live that could've been avoided, although I guess specifically more related to snapshots etc. rather than purely functionality. 

Sounds like @KrupaPammi's on the case anyways 👍

adamclark-dev commented:
@gusliedke makes sense to add tests at the component lib level. Now that component lib is going to be used in multiple applications, we don't want to have to be duplicating tests in multiple downstream applications.

tomisaacmarshall commented:
Thanks for comments all. Have made your requested changes @gusliedke 

Have also just updated snapshots and added a commit message with the 'feat:' prefix.

@KrupaPammi do we need to wait for your integration tests before merging this?

KrupaPammi commented:
> Thanks for comments all. Have made your requested changes @gusliedke
> 
> Have also just updated snapshots and added a commit message with the 'feat:' prefix.
> 
> @KrupaPammi do we need to wait for your integration tests before merging this?

@tomisaacmarshall please go ahead, I'm doing it on separate branch.

tomisaacmarshall commented:
> Looks really decent @tomisaacmarshall, nice one. Only addition it needs, comparing to the existing Storybook component, is to actual render out the chosen details underneath.. unless I'm missing something?

I was thinking the lookup input component itself should just do the lookup + have some means to provide the selected school data to its parent - i.e. the onSelect prop. Then it's up to the parent to decide what to do with that data e.g. print it out.

Then I was planning to do a follow up PR that would add the address form as another molecule, plus an organism tying everything together, which will be the thing that includes all the functionality of the old component (including rendering the address of the chosen school). Does that sounds ok?

AndyEPhipps commented:
> > Looks really decent @tomisaacmarshall, nice one. Only addition it needs, comparing to the existing Storybook component, is to actual render out the chosen details underneath.. unless I'm missing something?
> 
> I was thinking the lookup input component itself should just do the lookup + have some means to provide the selected school data to its parent - i.e. the onSelect prop. Then it's up to the parent to decide what to do with that data e.g. print it out.
> 
> Then I was planning to do a follow up PR that would add the manual address form (although it won't technically be a `<form>`...) as another molecule, plus an organism tying everything together, which will be the thing that includes all the functionality of the old component (including rendering the address of the chosen school). Does that sounds ok?

Okay, sweet; sounds like you've got it sussed, nice one 👍

gusliedke commented:
> @gusliedke makes sense to add tests at the component lib level. Now that component lib is going to be used in multiple applications, we don't want to have to be duplicating tests in multiple downstream applications.

That's true! 👍

AndyEPhipps commented:
@tomisaacmarshall : happy to let you add in those additional elements in separate issues and components if you feel like it's the right approach.

adamclark-dev commented:
:tada: This PR is included in version 3.21.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@comicrelief/component-library)
- [GitHub release](https://github.com/comicrelief/component-library/releases/tag/v3.21.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
